### PR TITLE
Fetch more keys -  review messages

### PR DESCRIPTION
**Pull request by sashaCher**
### Desired Outcome

All messages are introduced by adding `ca-cert` and `public-keys` variables are reviewed by TW.

### Connected Issue/Story

[ONYX-16152](https://ca-il-jira.il.cyber-ark.com:8443/browse/ONYX-16152)

#### Changelog

- [ ] The CHANGELOG has been updated, or
- [X] This PR does not include user-facing changes and doesn't require a
  CHANGELOG update

#### Test coverage

- [ ] This PR includes new unit and integration tests to go with the code
  changes, or
- [X] The changes in this PR do not require tests

#### Documentation

- [ ] Docs (e.g. `README`s) were updated in this PR
- [ ] A follow-up issue to update official docs has been filed here: [insert issue ID]()
- [X] This PR does not require updating any documentation

#### Behavior

- [ ] This PR changes product behavior and has been reviewed by a PO, or
- [ ] These changes are part of a larger initiative that will be reviewed later, or
- [X] No behavior was changed with this PR

#### Security

- [ ] Security architect has reviewed the changes in this PR,
- [ ] These changes are part of a larger initiative with a separate security review, or
- [X] There are no security aspects to these changes

### CODE REVIEW

File: app/domain/authentication/authn_jwt/signing_key/public_signing_keys.rb [hunk 1]
```diff
@@ -11,11 +11,11 @@ class PublicSigningKeys
         attr_required(:type, :value)
 
         validates(*required_attributes, presence: true)
-        validates(:type, inclusion: { in: VALID_TYPES, message: "'%{value}' is not a valid public-keys type" })
+        validates(:type, inclusion: { in: VALID_TYPES, message:  "'%{value}' is not a valid public-keys type" })
```
Comment (szamir1) [inline app/domain/authentication/authn_jwt/signing_key/public_signing_keys.rb:14]: can you pls add "valid value is jwks"
so the user will have the ability to quickly resolve the issue?

File: app/domain/errors.rb [hunk 2]
```diff
@@ -575,17 +575,17 @@ module AuthnJwt
...
 
       InvalidSigningKeyType = ::Util::TrackableErrorClass.new(
-        msg: "Signing key type '{0-type}' is invalid",
+        msg:  "Signing key type '{0-type}' is invalid",
         code: "CONJ00121E"
       )
 
       InvalidSigningKeySettings = ::Util::TrackableErrorClass.new(
-        msg: "Invalid signing key settings: {0-validation-error}",
+        msg:  "Invalid signing key settings: {0-validation-error}",
```
Comment (szamir1) [inline app/domain/errors.rb:588]: what is the difference from this one to CONJ00121E ?

**Review by szamir1 — CHANGES_REQUESTED** [review]

File: app/domain/errors.rb (hunk 2 above)
Comment (sashaCher) [reply app/domain/errors.rb:588]: CONJ00122E - is a validation time error. Will appear each time when configuration is invalid.
CONJ00121E - is a runtime error of the `CreateSigningKeyProvider` factory class. It exists in order to make the factory [switch case statement](https://github.com/cyberark/conjur/blob/3965075b5ff82da63cee6e0b3ba6fc5eecf09a9b/app/domain/authentication/authn_jwt/signing_key/create_signing_key_provider.rb#L50) fully deterministic. I guess that a customer will never see CONJ00121E error.

File: app/domain/authentication/authn_jwt/signing_key/public_signing_keys.rb (hunk 1 above)
Comment (sashaCher) [reply app/domain/authentication/authn_jwt/signing_key/public_signing_keys.rb]: Done. See now

**Review by szamir1 — APPROVED** [review]

**Comment by shulifink** [comment]
@sashaCher  - where do you want me to review?

File: CHANGELOG.md [hunk 3]
```diff
@@ -7,17 +7,17 @@ and this project adheres to [Semantic Versioning](http://semver.org/spec/v2.0.0.
 ## [Unreleased]
 
 ### Added
-- Added an ability to fetch signing keys from JWKS endpoints are using self-signed
-  certificate or certificate signed by 3rd party CA to JWT generic vendor configuration
+-  Added an ability to fetch signing keys from JWKS endpoints are using self-signed
```
Comment (shulifink) [inline CHANGELOG.md:10]:
```suggestion
-  Added the ability to fetch signing keys from JWKS endpoints that use a self-signed
```

File: CHANGELOG.md [hunk 4]
```diff
@@ -7,17 +7,17 @@ and this project adheres to [Semantic Versioning](http://semver.org/spec/v2.0.0.
...
+   certificate or certificate signed by 3rd party CA to JWT generic vendor configuration
```
Comment (shulifink) [inline CHANGELOG.md:11]:
```suggestion
   certificate or a certificate signed by a third-party CA for JWT generic vendor configuration
```

File: CHANGELOG.md [hunk 5]
```diff
@@ -7,17 +7,17 @@ and this project adheres to [Semantic Versioning](http://semver.org/spec/v2.0.0.
...
   ([#2462](https://github.com/cyberark/conjur/pull/2462)
   [#2461](https://github.com/cyberark/conjur/pull/2461)
   [#2456](https://github.com/cyberark/conjur/pull/2456)
   [#2455](https://github.com/cyberark/conjur/pull/2455)
   [#2457](https://github.com/cyberark/conjur/pull/2457)
   [#2452](https://github.com/cyberark/conjur/pull/2452)
   [#2437](https://github.com/cyberark/conjur/pull/2437))
-- Added an ability to JWT generic vendor configuration to receive signing keys for
-  JWT token verification from a variable. Variable name is `public-keys`
+-  Added an ability to JWT generic vendor configuration to receive signing keys for
```
Comment (shulifink) [inline CHANGELOG.md:19]:
```suggestion
-  Added the ability for JWT generic vendor configuration to receive signing keys for
```

File: app/domain/authentication/authn_jwt/signing_key/public_signing_keys.rb [hunk 6]
```diff
@@ -11,11 +11,11 @@ class PublicSigningKeys
         attr_required(:type, :value)
 
         validates(*required_attributes, presence: true)
-        validates(:type, inclusion: { in: VALID_TYPES, message: "'%{value}' is not a valid public-keys type" })
+        validates(:type, inclusion: { in: VALID_TYPES, message: "'%{value}' is not a valid public-keys type. Valid types are: #{VALID_TYPES.join(',')}" })
         validate(:validate_value_is_jwks, if: -> { @type == "jwks" })
 
         def initialize(hash)
-          raise Errors::Authentication::AuthnJwt::InvalidPublicKeys, "the value is not in valid JSON format" unless
+          raise Errors::Authentication::AuthnJwt::InvalidPublicKeys,  "the value is not in valid JSON format" unless
```
Comment (shulifink) [inline app/domain/authentication/authn_jwt/signing_key/public_signing_keys.rb:18]:
```suggestion
          raise Errors::Authentication::AuthnJwt::InvalidPublicKeys,  "Value not in valid JSON format" unless
```

File: app/domain/authentication/authn_jwt/signing_key/public_signing_keys.rb [hunk 7]
```diff
@@ -31,7 +31,7 @@ def validate!
         private
 
         def validate_value_is_jwks
-          errors.add(:value, "is not a valid JWKS (RFC7517)") unless @value.is_a?(Hash) &&
+          errors.add(:value,  "is not a valid JWKS (RFC7517)") unless @value.is_a?(Hash) &&
```
Comment (shulifink) [inline app/domain/authentication/authn_jwt/signing_key/public_signing_keys.rb:34]: fine

File: app/domain/authentication/authn_jwt/signing_key/signing_key_settings_builder.rb [hunk 8]
```diff
@@ -2,13 +2,13 @@ module Authentication
...
     module SigningKey
 
-      NO_SIGNING_KEYS_SOURCE = "One of #{JWKS_URI_RESOURCE_NAME}, #{PUBLIC_KEYS_RESOURCE_NAME}, and #{PROVIDER_URI_RESOURCE_NAME} have to be defined".freeze
-      ALL_SIGNING_KEYS_SOURCES = "#{JWKS_URI_RESOURCE_NAME}, #{PUBLIC_KEYS_RESOURCE_NAME}, and #{PROVIDER_URI_RESOURCE_NAME} cannot be define simultaneously".freeze
-      JWKS_PROVIDER_URI_SIGNING_PAIR = "#{JWKS_URI_RESOURCE_NAME} and #{PROVIDER_URI_RESOURCE_NAME} cannot be define simultaneously".freeze
-      JWKS_URI_PUBLIC_KEYS_PAIR = "#{JWKS_URI_RESOURCE_NAME} and #{PUBLIC_KEYS_RESOURCE_NAME} cannot be define simultaneously".freeze
-      PUBLIC_KEYS_PROVIDER_URI_PAIR = "#{PUBLIC_KEYS_RESOURCE_NAME} and #{PROVIDER_URI_RESOURCE_NAME} cannot be define simultaneously".freeze
-      CERT_STORE_ONLY_WITH_JWKS_URI = "#{CA_CERT_RESOURCE_NAME} can be define only with #{JWKS_URI_RESOURCE_NAME}".freeze
-      PUBLIC_KEYS_HAVE_ISSUER = "#{ISSUER_RESOURCE_NAME} is mandatory when #{PUBLIC_KEYS_RESOURCE_NAME} is set".freeze
+      NO_SIGNING_KEYS_SOURCE =  "One of #{JWKS_URI_RESOURCE_NAME}, #{PUBLIC_KEYS_RESOURCE_NAME}, and #{PROVIDER_URI_RESOURCE_NAME} have to be defined".freeze
```
Comment (shulifink) [inline app/domain/authentication/authn_jwt/signing_key/signing_key_settings_builder.rb:5]:
```suggestion
      NO_SIGNING_KEYS_SOURCE =  "One of the following must be defined: #{JWKS_URI_RESOURCE_NAME}, #{PUBLIC_KEYS_RESOURCE_NAME}, or #{PROVIDER_URI_RESOURCE_NAME}".freeze
```

File: app/domain/authentication/authn_jwt/signing_key/signing_key_settings_builder.rb [hunk 9]
```diff
@@ -2,13 +2,13 @@ module Authentication
...
+      ALL_SIGNING_KEYS_SOURCES =  "#{JWKS_URI_RESOURCE_NAME}, #{PUBLIC_KEYS_RESOURCE_NAME}, and #{PROVIDER_URI_RESOURCE_NAME} cannot be define simultaneously".freeze
```
Comment (shulifink) [inline app/domain/authentication/authn_jwt/signing_key/signing_key_settings_builder.rb:6]:
```suggestion
      ALL_SIGNING_KEYS_SOURCES =  "#{JWKS_URI_RESOURCE_NAME}, #{PUBLIC_KEYS_RESOURCE_NAME}, and #{PROVIDER_URI_RESOURCE_NAME} cannot be defined simultaneously".freeze
```

File: app/domain/authentication/authn_jwt/signing_key/signing_key_settings_builder.rb [hunk 10]
```diff
@@ -2,13 +2,13 @@ module Authentication
...
+      JWKS_PROVIDER_URI_SIGNING_PAIR =  "#{JWKS_URI_RESOURCE_NAME} and #{PROVIDER_URI_RESOURCE_NAME} cannot be define simultaneously".freeze
```
Comment (shulifink) [inline app/domain/authentication/authn_jwt/signing_key/signing_key_settings_builder.rb:7]:
```suggestion
      JWKS_PROVIDER_URI_SIGNING_PAIR =  "#{JWKS_URI_RESOURCE_NAME} and #{PROVIDER_URI_RESOURCE_NAME} cannot be defined simultaneously".freeze
```

File: app/domain/authentication/authn_jwt/signing_key/signing_key_settings_builder.rb [hunk 11]
```diff
@@ -2,13 +2,13 @@ module Authentication
...
+      JWKS_URI_PUBLIC_KEYS_PAIR =  "#{JWKS_URI_RESOURCE_NAME} and #{PUBLIC_KEYS_RESOURCE_NAME} cannot be define simultaneously".freeze
```
Comment (shulifink) [inline app/domain/authentication/authn_jwt/signing_key/signing_key_settings_builder.rb:8]:
```suggestion
      JWKS_URI_PUBLIC_KEYS_PAIR =  "#{JWKS_URI_RESOURCE_NAME} and #{PUBLIC_KEYS_RESOURCE_NAME} cannot be defined simultaneously".freeze
```

File: app/domain/authentication/authn_jwt/signing_key/signing_key_settings_builder.rb [hunk 12]
```diff
@@ -2,13 +2,13 @@ module Authentication
...
+      PUBLIC_KEYS_PROVIDER_URI_PAIR =  "#{PUBLIC_KEYS_RESOURCE_NAME} and #{PROVIDER_URI_RESOURCE_NAME} cannot be define simultaneously".freeze
```
Comment (shulifink) [inline app/domain/authentication/authn_jwt/signing_key/signing_key_settings_builder.rb:9]:
```suggestion
      PUBLIC_KEYS_PROVIDER_URI_PAIR =  "#{PUBLIC_KEYS_RESOURCE_NAME} and #{PROVIDER_URI_RESOURCE_NAME} cannot be defined simultaneously".freeze
```

File: app/domain/authentication/authn_jwt/signing_key/signing_key_settings_builder.rb [hunk 13]
```diff
@@ -2,13 +2,13 @@ module Authentication
...
+      CERT_STORE_ONLY_WITH_JWKS_URI =  "#{CA_CERT_RESOURCE_NAME} can be define only with #{JWKS_URI_RESOURCE_NAME}".freeze
```
Comment (shulifink) [inline app/domain/authentication/authn_jwt/signing_key/signing_key_settings_builder.rb:10]:
```suggestion
      CERT_STORE_ONLY_WITH_JWKS_URI =  "#{CA_CERT_RESOURCE_NAME} can only be defined together with #{JWKS_URI_RESOURCE_NAME}".freeze
```

File: app/domain/authentication/authn_jwt/signing_key/signing_key_settings_builder.rb [hunk 14]
```diff
@@ -2,13 +2,13 @@ module Authentication
...
+      PUBLIC_KEYS_HAVE_ISSUER =  "#{ISSUER_RESOURCE_NAME} is mandatory when #{PUBLIC_KEYS_RESOURCE_NAME} is set".freeze
```
Comment (shulifink) [inline app/domain/authentication/authn_jwt/signing_key/signing_key_settings_builder.rb:11]:
```suggestion
      PUBLIC_KEYS_HAVE_ISSUER =  "#{ISSUER_RESOURCE_NAME} is mandatory when #{PUBLIC_KEYS_RESOURCE_NAME} is defined".freeze
```

File: app/domain/errors.rb [hunk 15]
```diff
@@ -575,17 +575,17 @@ module AuthnJwt
       )
 
       InvalidPublicKeys = ::Util::TrackableErrorClass.new(
-        msg: "Failed to parse 'public-keys': {0-parse-error}",
+        msg:  "Failed to parse 'public-keys': {0-parse-error}",
```
[the remark this answers is not in the document]
Comment (shulifink) [reply app/domain/errors.rb:578]: fine

File: app/domain/errors.rb [hunk 16]
```diff
@@ -575,17 +575,17 @@ module AuthnJwt
...
         code: "CONJ00120E"
       )
 
       InvalidSigningKeyType = ::Util::TrackableErrorClass.new(
-        msg: "Signing key type '{0-type}' is invalid",
+        msg:  "Signing key type '{0-type}' is invalid",
```
[the remark this answers is not in the document]
Comment (shulifink) [reply app/domain/errors.rb:583]: fine

File: app/domain/logs.rb [hunk 17]
```diff
@@ -720,12 +720,12 @@ module AuthnJwt
       )
 
       ParsingStaticSigningKeys = ::Util::TrackableLogMessageClass.new(
-        msg: "Parsing JWKS from public-keys value",
+        msg:  "Parsing JWKS from public-keys value",
```
Comment (shulifink) [inline app/domain/logs.rb:723]:
```suggestion
        msg:  "Parsing JWKS from public-keys value...",
```

File: app/domain/errors.rb (hunk 2 above)
Comment (shulifink) [reply app/domain/errors.rb:588]: fine from English point of view

File: app/domain/logs.rb [hunk 18]
```diff
@@ -720,12 +720,12 @@ module AuthnJwt
...
         code: "CONJ00143I"
       )
 
       ParsedStaticSigningKeys = ::Util::TrackableLogMessageClass.new(
-        msg: "Successfully parsed public-keys value",
+        msg:  "Successfully parsed public-keys value",
```
[the remark this answers is not in the document]
Comment (shulifink) [reply app/domain/logs.rb:728]: fine

**Review by shulifink — CHANGES_REQUESTED** [review]

**Review by shulifink — APPROVED** [review]

**Review by szamir1 — DISMISSED** [review]

**Review by szamir1 — APPROVED** [review]

**Pull request merged**